File: app/src/main/java/com/elvotra/clean/domain/usecase/GetPostUseCase.java
package com.elvotra.clean.domain.usecase;

import com.elvotra.clean.domain.model.Post;
import com.elvotra.clean.domain.repository.IPostsRepository;
import com.elvotra.clean.domain.usecase.base.BaseUseCase;

public class GetPostUseCase extends BaseUseCase<GetPostUseCase.RequestValues, GetPostUseCase.ResponseValue> {
    private IPostsRepository repository;

    public GetPostUseCase(IPostsRepository repository) {
        this.repository = repository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        int postId = requestValues.getPostId();
        repository.getPost(postId, new IPostsRepository.LoadPostCallback() {
            @Override
            public void onPostLoaded(Post post) {
                GetPostUseCase.ResponseValue responseValue = new GetPostUseCase.ResponseValue(post);
                getUseCaseCallback().onSuccess(responseValue);
            }

            @Override
            public void onError(final int statusCode) {
                getUseCaseCallback().onError(statusCode);
            }
        });
    }

    public static final class RequestValues implements BaseUseCase.RequestValues {

        private final int postId;

        public RequestValues(int postId) {
            this.postId = postId;
        }

        public int getPostId() {
            return postId;
        }
    }

    public static final class ResponseValue implements BaseUseCase.ResponseValue {

        private Post post;

        public ResponseValue(Post post) {
            this.post = post;
        }

        public Post getPost() {
            return this.post;
        }
    }
}

File: app/src/main/java/com/elvotra/clean/domain/usecase/GetPostUseCase.kt
package com.elvotra.clean.domain.usecase

import com.elvotra.clean.domain.model.Post
import com.elvotra.clean.domain.repository.IPostsRepository
import com.elvotra.clean.domain.usecase.base.BaseUseCase

class GetPostUseCase(private val repository: IPostsRepository) : BaseUseCase<GetPostUseCase.RequestValues, GetPostUseCase.ResponseValue>() {

    override fun executeUseCase(requestValues: RequestValues?) {
        val postId = requestValues!!.postId
        repository.getPost(postId, object : IPostsRepository.LoadPostCallback {
            override fun onPostLoaded(post: Post) {
                val responseValue = GetPostUseCase.ResponseValue(post)
                useCaseCallback!!.onSuccess(responseValue)
            }

            override fun onError(statusCode: Int) {
                useCaseCallback!!.onError(statusCode)
            }
        })
    }

    class RequestValues(val postId: Int) : BaseUseCase.RequestValues

    class ResponseValue(val post: Post) : BaseUseCase.ResponseValue
}
